ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs

```diff
@@ -0,0 +1,320 @@
+import           Cardano.Ledger.Shelley.API
+                     (NewEpochState (stashedAVVMAddresses), ShelleyGenesis (..),
+                     ShelleyGenesisStaking (..), TxIn (..),
+                     translateCompactTxOutByronToShelley,
+                     translateTxIdByronToShelley)
+import           Cardano.Ledger.Shelley.PParams (emptyPParams)
+import           Cardano.Ledger.Shelley.UTxO
+
+import           Ouroboros.Consensus.BlockchainTime.WallClock.Types
+                     (slotLengthFromSec)
+import           Ouroboros.Consensus.Cardano.Block (CardanoEras)
+import           Ouroboros.Consensus.Cardano.CanHardFork ()
+import           Ouroboros.Consensus.HardFork.Combinator (InPairs (..),
+                     hardForkEraTranslation, translateLedgerState)
+import           Ouroboros.Consensus.HardFork.Combinator.State.Types
+                     (TranslateLedgerState (translateLedgerStateWith))
+import           Ouroboros.Consensus.HardFork.Combinator.Util.InPairs (AllPairs,
+                     RequiringBoth, provideBoth)
+import           Ouroboros.Consensus.Ledger.Basics (ApplyMapKind' (..), DiffMK,
+                     EmptyMK, LedgerConfig, LedgerState)
+import           Ouroboros.Consensus.Protocol.Praos
+import           Ouroboros.Consensus.Protocol.Praos.Common
+                     (MaxMajorProtVer (..))
+import           Ouroboros.Consensus.Protocol.Praos.Translate ()
+import           Ouroboros.Consensus.Protocol.TPraos (TPraos)
+import           Ouroboros.Consensus.TypeFamilyWrappers
+import           Ouroboros.Consensus.Util (dimap)
+
+import           Ouroboros.Consensus.Byron.Ledger (ByronBlock, byronLedgerState)
+
+import           Ouroboros.Consensus.Shelley.Eras
+import           Ouroboros.Consensus.Shelley.HFEras ()
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock,
+                     ShelleyLedgerConfig, mkShelleyLedgerConfig,
+                     shelleyLedgerState, shelleyLedgerTables, shelleyUTxOTable)
+import           Ouroboros.Consensus.Shelley.Ledger.SupportsProtocol ()
+
+-- TODO: Not sure where to put these
+import           Test.Cardano.Ledger.Alonzo.Serialisation.Generators ()
+import           Test.Cardano.Ledger.Shelley.Utils (unsafeBoundRational)
+import           Test.Consensus.Byron.Generators (genByronLedgerConfig,
+                     genByronLedgerState)
+import           Test.Consensus.Cardano.MockCrypto (MockCryptoCompatByron)
+import           Test.Consensus.Shelley.Generators ()
+import           Test.Consensus.Shelley.MockCrypto
+
+import           Test.Util.Time (dawnOfTime)
+
+-- Definitions to make the signatures a bit less unwieldy
+type Crypto = MockCryptoCompatByron
+type Proto  = TPraos Crypto
+
+tests :: TestTree
+tests = testGroup "UpdateTablesOnEraTransition" $ toTestTrees predicates tls
+  where
+    tls = translateLedgerState hardForkEraTranslation
+    predicates :: InPairs TranslationPredicate (CardanoEras Crypto)
```

Comment:
Why not define a `[TestTree]` directly? Is `toTestTrees` necessary?

Reply:
Well, I couldn't do it.

Reply:
To define the properties I need to have access to the appropriate `TranslateEra x y` values. I chose the direction of zipping the predicates with what is loosely, `InPairs TranslateEra (CardanoEras c)`.